framework/src/org/checkerframework/framework/util/LubVisitor.java

```diff
@@ -0,0 +1,350 @@
+package org.checkerframework.framework.util;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.type.TypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedIntersectionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedNullType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedPrimitiveType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedUnionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor;
+import org.checkerframework.framework.type.visitor.VisitHistory;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+public class LubVisitor extends AbstractAtmComboVisitor<Void, AnnotatedTypeMirror> {
+
+    public AnnotatedTypeFactory atypeFactory;
+    public QualifierHierarchy qualifierHierarchy;
+    public VisitHistory visitHistory = new VisitHistory();
+
+    public LubVisitor(AnnotatedTypeFactory atypeFactory) {
+        this.atypeFactory = atypeFactory;
+        this.qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+    }
+
+    /**
+     * Returns an ATM that is the least upper bound of type1 and type2 and whose Java type is
+     * lubJavaType.  lubJavaType must be a super type or convertible to the Java types of type1 and
+     * type2.
+     **/
+    AnnotatedTypeMirror lub(
```

Comment:
This seems to be the main entry point for this class. It seems inconsistent to have the class and constructor public, but this method package private. I guess they should all be package private with a comment about who is supposed to use this class.